### PR TITLE
Open lambda

### PR DESCRIPTION
**Pull request by NathanTP**
Open Lambda and AWS now both work with basic 'package', 'install', 'create', and 'bench' commands. This also subsumes PR #3 so I will close that. This is a sufficiently working set of abstractions for FaaS and a tentative starting point for builtin benchmarks (although the concurrency sweep still doesn't work).

Documentation and better logging will be managed in a separate branch.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,9 +27,8 @@ This will create the srk binary locally. To install to your GOPATH:
 Create a zip file containing the example workload:
```
Comment (jssmith) [inline README.md]: Add simple echo example

File: cmd/bench.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	"net"
+
+	"github.com/serverlessresearch/srk/pkg/cfbench"
+	"github.com/serverlessresearch/srk/pkg/srk"
+	"github.com/spf13/cobra"
+)
+
+// Filled in by cobra argument parsing in init()
+var benchCmdConfig struct {
+	bench        string
```
Comment (jssmith) [inline cmd/bench.go]: benchmarkName

File: cmd/bench.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+		switch benchCmdConfig.bench {
+		case "oneShot":
+			var err error
+			bench, err = cfbench.NewOneShot()
+			if err != nil {
+				panic("Failed to initialize OneShot benchmark")
+			}
+
+		case "concurrency_scan":
```
Comment (jssmith) [inline cmd/bench.go]: Need to decide on camelCase vs under_scores

File: cmd/config.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	awslambda "github.com/serverlessresearch/srk/pkg/aws-lambda"
+	"github.com/serverlessresearch/srk/pkg/openlambda"
+	"github.com/serverlessresearch/srk/pkg/srk"
+	"github.com/spf13/viper"
+)
+
+func getProvider() *srk.Provider {
+	return &srk.Provider{getFaasService()}
```
Comment (jssmith) [inline cmd/config.go]: "FaaS Service" is redundant

File: cmd/install.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/*
+Copyright © 2019 NAME HERE <EMAIL ADDRESS>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
```
Comment (jssmith) [inline cmd/install.go]: Delete license boilerplate

File: cmd/install.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+// installCmd represents the install command
+var installCmd = &cobra.Command{
+	Use:   "install",
+	Short: "Install a pre-packaged function to the configured FaaS service",
+	Long:  `Install a function to the FaaS service. It is assumed that you have already packaged this function (using the 'package' command).`,
+	Run: func(cmd *cobra.Command, args []string) {
+		service := getFaasService()
+		defer service.Destroy()
+
+		rawDir := getRawPath(installName)
```
Comment (jssmith) [inline cmd/install.go]: Consider making this an optional name that defaults to the name of a folder.

File: cmd/package.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		packageCmdConfig.service = getFaasService()
+		defer packageCmdConfig.service.Destroy()
+
+		funcName := strings.TrimSuffix(path.Base(packageCmdConfig.source), path.Ext(packageCmdConfig.source))
+		includes := strings.Split(packageCmdConfig.include, ",")
+		rawDir := getRawPath(funcName)
+
+		if err := createRaw(packageCmdConfig.source, funcName, includes, rawDir); err != nil {
+			fmt.Println("Packaging function failed: %v\n", err)
+			return
```
Comment (jssmith) [inline cmd/package.go]: Does Cobra provide a way to provide a status code on return?

File: cmd/root.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	// but aren't included by default.
+
+	// Dumping ground for all generated output. Users should be able to "rm -rf
+	// build" and get a clean system.
+	viper.SetDefault("buildDir", "./build")
+
+	// Collects all srk-provided libraries.
+	viper.SetDefault("includeDir", "./includes")
+
+	viper.SetDefault("service.faas.awsLambda.region", "us-west-2")
```
Comment (jssmith) [inline cmd/root.go]: Can we just rely on the environment variable `AWS_DEFAULT_REGION`?

File: configs/srk.yaml [hunk 9]
```diff
@@ -0,0 +1,52 @@
+# This provider will be used by srk commands unless specifically overwritten
```
Comment (jssmith) [inline configs/srk.yaml]: make a template file for now, let people copy it over and modify the fields that they need.

File: examples/cfbench/f.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# This is a compatibility shim with open-lambda. OL requires that
```
Comment (jssmith) [inline examples/cfbench/f.py]: Update comment to reflect intention to change this in OpenLambda.

File: pkg/openlambda/ol.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+func (self *olConfig) Destroy() {
+	if self.sessionStarted {
+		self.terminateOlWorker()
+	}
+}
+
+func (self *olConfig) Invoke(fName string, args string) (resp *bytes.Buffer, rerr error) {
+	if !self.sessionStarted {
+		self.launchOlWorker()
```
Comment (jssmith) [inline pkg/openlambda/ol.go]: We might want a separate command to launch the worker / boot up the OpenLambda environment. Consider how it works with AWS - Lambda is running in the background all the time and maintains cached function instances. The invoke command just runs a function.

File: pkg/srk/interface.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	Install(rawDir string) (rerr error)
+
+	// Invoke function
+	// fName: Name of function
+	// args: JSON-encoded argument string
+	// Returns: function response as a bytes buffer. The exact format of this
+	// response may depend on the FaaS service. resp may be nil (indicating no
+	// valid response was received)
+	Invoke(fName string, args string) (resp *bytes.Buffer, rerr error)
+
```
Comment (jssmith) [inline pkg/srk/interface.go]: Add Remove / Delete call

File: cmd/bench.go (hunk 3 above)
Comment (NathanTP) [reply cmd/bench.go]: Now that i think about it (and googling around a bit), I think dashes are pretty standard for cli args. I'll go with dashes, so "one-shot" and "concurrency-scan". Internally, we'll stick with CamelCase for names of things in code (I'm not sure if golang or others have an opinion on that, but I've been using CamelCase).

File: pkg/srk/interface.go (hunk 12 above)
Comment (NathanTP) [reply pkg/srk/interface.go]: I added a "remove" command that removes the function from the service (i.e. the opposite of Install). I also added a "clean" command that cleans up local working files (everything by default, or just one function if provided with a name parameter).

**Pull request merged**